configuration/config2/XSDSchemaParser.cpp

```diff
@@ -719,22 +719,63 @@ void XSDSchemaParser::parseAllowedValue(const pt::ptree &allowedValueTree, Schem
 std::shared_ptr<SchemaValue> XSDSchemaParser::getSchemaValue(const pt::ptree &attr)
 {
     std::string attrName = getXSDAttributeValue(attr, "<xmlattr>.name");
+
+    if (!attr.get("<xmlattr>.default", "").empty())
+    {
+        throw(ParseException( "Attribute " + m_pSchemaItem->getProperty("name") + "[@" + attrName + "], XSD default is not supported, use hpcc:presetValue or hpcc:forcedConfigValue instead"));
+    }
+
     std::shared_ptr<SchemaValue> pCfgValue = std::make_shared<SchemaValue>(attrName);
     pCfgValue->setDisplayName(attr.get("<xmlattr>.hpcc:displayName", attrName));
     pCfgValue->setRequired(attr.get("<xmlattr>.use", "optional") == "required");
     pCfgValue->setTooltip(attr.get("<xmlattr>.hpcc:tooltip", ""));
     pCfgValue->setReadOnly(attr.get("<xmlattr>.hpcc:readOnly", "false") == "true");
-    pCfgValue->setHidden(attr.get("<xmlattr>.hpcc:hidden", "false") == "true");
     pCfgValue->setDeprecated(attr.get("<xmlattr>.hpcc:deprecated", "false") == "true");
     pCfgValue->setMirrorFromPath(attr.get("<xmlattr>.hpcc:mirrorFrom", ""));
     pCfgValue->setAutoGenerateType(attr.get("<xmlattr>.hpcc:autoGenerateType", ""));
     pCfgValue->setAutoGenerateValue(attr.get("<xmlattr>.hpcc:autoGenerateValue", ""));
-    pCfgValue->setDefaultValue(attr.get("<xmlattr>.default", ""));
-    pCfgValue->setCodeDefault(attr.get("<xmlattr>.hpcc:defaultInCode", ""));
     pCfgValue->setValueLimitRuleType(attr.get("<xmlattr>.hpcc:valueLimitRuleType", ""));
     pCfgValue->setValueLimitRuleData(attr.get("<xmlattr>.hpcc:valueLimitRuleData", ""));
     pCfgValue->setRequiredIf(attr.get("<xmlattr>.hpcc:requiredIf", ""));
 
+    //
+    // Process the various hidden/visible flags and ensure no conflicts
+    std::string hidden = attr.get("<xmlattr>.hpcc:hidden", "");
+    std::string hiddenIf = attr.get("<xmlattr>.hpcc:hiddenIf", "");
+    std::string visibleIf = attr.get("<xmlattr>.hpcc:visibleIf", "");
+    unsigned countAttrs = (hidden.empty() ? 0 : 1) + (hiddenIf.empty() ? 0 : 1) + (visibleIf.empty() ? 0 : 1);
+    if (countAttrs > 1)
+    {
+        throw(ParseException( "Attribute " + m_pSchemaItem->getProperty("name") + "[@" + attrName + "] Only one of hpcc:hidden, hpcc:hiddenIf or hpcc:visibleIf may be specified"));
+    }
+
+    if (!hidden.empty())
+    {
+        pCfgValue->setHidden(hidden == "true");
+    }
+    else
+    {
+        pCfgValue->setHiddenIf(hiddenIf + visibleIf);
```

Comment:
what is actually being done here? 
we know one of hiddenIf or visibleIf is empty, but why are you appending hiddenif and visibleIf ?

Reply:
this is going to cause maintainability problems, let's rewrite using verbose if else cluases

Reply:
Changed to use a ternary operator so it's clear which path attribute is being used.